In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)

base_temp = 10.0
amplitude = 14
city = "Дніпро"

rows = []

for year in [2021, 2022, 2023, 2024]:
    for month in range(1, 13):
        seasonal = amplitude * np.cos((month - 7) / 12 * 2 * np.pi)
        noise = np.random.normal(0, 1.0)

        rows.append({
            "місто": city,
            "рік": year,
            "місяць": month,
            "температура": round(base_temp + seasonal + noise, 1)
        })

climate = pd.DataFrame(rows)

print("Варіант 5 — Дніпро")
print(climate)

Варіант 5 — Дніпро
     місто   рік  місяць  температура
0   Дніпро  2021       1         -3.5
1   Дніпро  2021       2         -2.3
2   Дніпро  2021       3          3.6
3   Дніпро  2021       4         11.5
4   Дніпро  2021       5         16.8
5   Дніпро  2021       6         21.9
6   Дніпро  2021       7         25.6
7   Дніпро  2021       8         22.9
8   Дніпро  2021       9         16.5
9   Дніпро  2021      10         10.5
10  Дніпро  2021      11          2.5
11  Дніпро  2021      12         -2.6
12  Дніпро  2022       1         -3.8
13  Дніпро  2022       2         -4.0
14  Дніпро  2022       3          1.3
15  Дніпро  2022       4          9.4
16  Дніпро  2022       5         16.0
17  Дніпро  2022       6         22.4
18  Дніпро  2022       7         23.1
19  Дніпро  2022       8         20.7
20  Дніпро  2022       9         18.5
21  Дніпро  2022      10          9.8
22  Дніпро  2022      11          3.1
23  Дніпро  2022      12         -3.5
24  Дніпро  2023       1       

In [2]:
year_stats = climate.groupby("рік")["температура"].agg(
    ["mean", "min", "max"]
)

print(year_stats)

           mean  min   max
рік                       
2021  10.283333 -3.5  25.6
2022   9.416667 -4.0  23.1
2023   9.808333 -4.5  24.0
2024   9.658333 -4.1  23.9


Середня температура:
2021 — 10.28 °C
2022 — 9.42 °C
2023 — 9.81 °C
2024 — 9.66 °C.

Чіткого тренду до потепління або похолодання за ці чотири роки
не спостерігається. Температура коливається від року до року,
а зміни виглядають випадковими.

In [3]:
month_stats = climate.groupby("місяць")["температура"].agg(
    ["mean", "std"]
)

print(month_stats)

max_std_month = month_stats["std"].idxmax()
max_std = month_stats["std"].max()

print("\nНайбільший розкид:")
print("Місяць:", max_std_month)
print("std:", max_std)

          mean       std
місяць                  
1       -3.900  0.424264
2       -3.100  1.104536
3        2.100  1.023067
4       10.375  0.865544
5       16.725  0.727438
6       22.100  0.294392
7       24.000  1.116542
8       22.350  1.420094
9       16.875  1.250000
10       9.625  0.689807
11       2.975  0.618466
12      -2.625  1.087428

Найбільший розкид:
Місяць: 8
std: 1.420093893609386


Найбільший розкид температури між роками спостерігається
у серпні (місяць 8).

Стандартне відхилення становить приблизно 1.42 °C.

Це означає, що температура серпня в нашому наборі найбільше
відрізнялася між 2021–2024 роками. Оскільки дані синтетичні
та містять випадковий шум, така нестабільність може бути
результатом випадкових коливань.

In [4]:
pivot_temp = climate.pivot_table(
    index="місяць",
    columns="рік",
    values="температура",
    aggfunc="mean"
)

print(pivot_temp)

рік     2021  2022  2023  2024
місяць                        
1       -3.5  -3.8  -4.5  -3.8
2       -2.3  -4.0  -2.0  -4.1
3        3.6   1.3   1.8   1.7
4       11.5   9.4  10.4  10.2
5       16.8  16.0  16.4  17.7
6       21.9  22.4  21.8  22.3
7       25.6  23.1  23.4  23.9
8       22.9  20.7  24.0  21.8
9       16.5  18.5  17.0  15.5
10      10.5   9.8   8.9   9.3
11       2.5   3.1   3.8   2.5
12      -2.6  -3.5  -3.3  -1.1


Зведена таблиця pivot_table зручніша для читання людиною,
оскільки можна легко порівнювати температуру одного місяця
між різними роками.

Довгий (tidy) формат climate зручніший для аналізу даних,
групування, фільтрації та побудови графіків.

Отже, зведена форма зручніша для візуального порівняння,
а довгий формат — для подальшої обробки даних.

In [5]:
def get_season(month):
    if month in [12, 1, 2]:
        return "зима"
    elif month in [3, 4, 5]:
        return "весна"
    elif month in [6, 7, 8]:
        return "літо"
    else:
        return "осінь"

climate["сезон"] = climate["місяць"].apply(get_season)

climate["тепліше_за_середнє"] = (
    climate["температура"] > base_temp
)

print(climate.head(15))

     місто   рік  місяць  температура  сезон  тепліше_за_середнє
0   Дніпро  2021       1         -3.5   зима               False
1   Дніпро  2021       2         -2.3   зима               False
2   Дніпро  2021       3          3.6  весна               False
3   Дніпро  2021       4         11.5  весна                True
4   Дніпро  2021       5         16.8  весна                True
5   Дніпро  2021       6         21.9   літо                True
6   Дніпро  2021       7         25.6   літо                True
7   Дніпро  2021       8         22.9   літо                True
8   Дніпро  2021       9         16.5  осінь                True
9   Дніпро  2021      10         10.5  осінь                True
10  Дніпро  2021      11          2.5  осінь               False
11  Дніпро  2021      12         -2.6   зима               False
12  Дніпро  2022       1         -3.8   зима               False
13  Дніпро  2022       2         -4.0   зима               False
14  Дніпро  2022       3 

In [6]:
season_table = pd.crosstab(
    climate["сезон"],
    climate["тепліше_за_середнє"]
)

print(season_table)

тепліше_за_середнє  False  True 
сезон                           
весна                   5      7
зима                   12      0
літо                    0     12
осінь                   7      5


Результат crosstab відповідає очікуванням.

Усі 12 літніх спостережень мають температуру вище
середньорічної температури 10 °C.

Усі 12 зимових спостережень, навпаки, мають температуру
нижче або рівну цьому рівню.

Навесні та восени спостерігаються обидві категорії,
оскільки це перехідні сезони.

In [7]:
pivot_simple = climate.pivot(
    index="місяць",
    columns="рік",
    values="температура"
)

print(pivot_simple)

рік     2021  2022  2023  2024
місяць                        
1       -3.5  -3.8  -4.5  -3.8
2       -2.3  -4.0  -2.0  -4.1
3        3.6   1.3   1.8   1.7
4       11.5   9.4  10.4  10.2
5       16.8  16.0  16.4  17.7
6       21.9  22.4  21.8  22.3
7       25.6  23.1  23.4  23.9
8       22.9  20.7  24.0  21.8
9       16.5  18.5  17.0  15.5
10      10.5   9.8   8.9   9.3
11       2.5   3.1   3.8   2.5
12      -2.6  -3.5  -3.3  -1.1


Метод pivot() успішно працює для набору climate.

Причина полягає в тому, що для кожної комбінації
"місяць + рік" у наборі існує рівно один рядок.
Тому pivot() однозначно знає, яке значення температури
потрібно помістити в кожну клітинку.

Якби для одного міста було декілька метеостанцій і кожна
станція мала окреме вимірювання температури за той самий
місяць і рік, виникли б дублікати комбінацій "місяць + рік".

У такому випадку pivot() завершився б помилкою, оскільки
не знав би, яке з декількох значень вибрати.

Для таких даних можна використати pivot_table() з функцією
агрегації, наприклад aggfunc="mean".

Метод pivot() успішно працює для набору climate.

Причина полягає в тому, що для кожної комбінації
"місяць + рік" у наборі існує рівно один рядок.
Тому pivot() однозначно знає, яке значення температури
потрібно помістити в кожну клітинку.

Якби для одного міста було декілька метеостанцій і кожна
станція мала окреме вимірювання температури за той самий
місяць і рік, виникли б дублікати комбінацій "місяць + рік".

У такому випадку pivot() завершився б помилкою, оскільки
не знав би, яке з декількох значень вибрати.

Для таких даних можна використати pivot_table() з функцією
агрегації, наприклад aggfunc="mean".